### PR TITLE
Fabric avalon API 1.1 implementation

### PR DESCRIPTION
**Pull request by litong01**
Signed-off-by: Tong Li <litong01@us.ibm.com>

### CODE REVIEW

File: client_sdk/avalon_client_sdk/fabric/eventlistener.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/python
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/fabric/eventlistener.py]: Add license header

File: client_sdk/avalon_client_sdk/fabric/txcommiter.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/python
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/fabric/txcommiter.py]: Add license header

File: client_sdk/avalon_client_sdk/fabric/txcommiter.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        loop = asyncio.get_event_loop()
+        try:
+            responses = loop.run_until_complete(self.client.chaincode_query(
+                requestor=self.user, channel_name=self.channel_name,
+                peers=[self.peer_name], args=args, cc_name=cc_name))
+            logger.info('Tx response: {0}'.format(responses))
+            return responses
+        except Exception as ex:
+            logger.error('Query error: {0}'.format(ex))
+            return {}
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/fabric/txcommiter.py]: Add new line.

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/go.mod [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// limitations under the License.
+//
+
+module order
+
+go 1.12
+
+require (
+	github.com/golang/protobuf v1.3.1
+	github.com/hyperledger/fabric v1.4.1
```
Comment (Ram-srini) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/go.mod]: Still i could see v1,4.1 here, Does it support 2.0?

**Comment by Ram-srini** [comment]
@litong01 Looks like last commit doesn't the sign-off tag, DCO check is failing. Please do fix that.

**Comment by danintel** [comment]
You have a DCO sign-off on 2 of your 3 commits.  The 3rd commit is a merge commit without a sign-off and that is failing.  You should be able to rebase your 3 commits into 1 commit and force push it with git.
Something like:
`git rebase -i HEAD~3; git push -f origin master` # where origin = https://github.com/litong01/avalon-1

Also in the future avoid doing merges.  Instead rebase changes from upstream.  For example:
`it fetch upstream master; git rebase -i upstream/master`

**Comment by litong01** [comment]
squash commits. please review. Thanks.

File: client_sdk/avalon_client_sdk/fabric/eventlistener.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        try:
+            with open(self._config, 'r') as file:
+                data = file.read()
+                if len(data) > 0:
+                    self._last_block = int(data)
+            logger.info('Loaded block mark: %s', self._last_block)
+        except IOError:
+            pass
+
+    def saveConfig(self):
```
[the remark this answers is not in the document]
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/eventlistener.py]: hmmm. there are quite few names. Can it all be changed in the next pr?

File: client_sdk/avalon_client_sdk/fabric/eventlistener.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import asyncio
+import logging
+import grpc
+from hfc.fabric.peer import create_peer
+
+from avalon import base
+
+logger = logging.getLogger(__name__)
+
+class EventListener(base.ClientBase):
```
[the remark this answers is not in the document]
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/eventlistener.py]: Can it all be changed in the next pr?

File: client_sdk/avalon_client_sdk/fabric/txcommiter.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/python
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/txcommiter.py]: Change the name of the file from `txcommiter.py` to `txcommitter.py`

File: client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/main.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+// parameter. The registry must match to all input parameters (AND mode) to be
+// included in the list.
+// params:
+//    bytes32 workerServiceId,
+//    bytes32 workerId,
+//    bytes32 requesterId,
+//    uint256 receiptStatus
+// returns:
+//   int totalCount
+//   string LookupTag
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/main.go]: Change `LookupTag` to `lookupTag`
(see line 46)

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+package main
+
+// Worker related constants
+const (
+	WORKERACTIVE         = 1
+	WORKEROFFLINE        = 2
+	WORKERDECOMMISSIONED = 3
+	WORKERCOMPROMISED    = 4
+
+	OBJECTTYPE   = "WorkerRegister"
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go]: Change `WorkerRegister` to `WorkerRegistery`

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	PAGESIZE     = 10
+	UINT64FORMAT = "%020d"
+	BYTE32FORMAT = "%-32v"
+)
+
+// WorkerRegistry workerRegister object saved
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go]: Change `workerRegister` to `workerRegistery`

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+	return cert.Subject.CommonName
+}
+
+// Init the init function of the chaincode
+func (t *WorkOrder) Init(stub shim.ChaincodeStubInterface) pb.Response {
+	logger.Info("WorkOrder Init")
+	return shim.Success(nil)
+}
+
+// workOrderSubmit - This function submit a work order
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `submit` to `submits`

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+	return &param, nil
+}
+
+// getMSPID - This function retrieve the MSPID of the calling agent
+func (t *WorkOrder) getMSPID(stub shim.ChaincodeStubInterface) string {
+	var mspid string
+	id, err := cid.New(stub)
+	if err != nil {
+		logger.Errorf("Can not get peer id")
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `Can not` to `Cannot` here (line 59) and on lines 65, 73, and 86 below.

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+	err = stub.SetEvent("workOrderSubmitted", eventPayload)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	logger.Info("Finished workOrderSubmit")
+	return shim.Success(nil)
+}
+
+// workOrderComplete - This function submit a work order
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `submit` to `submits`

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+func (t *WorkOrder) workOrderComplete(stub shim.ChaincodeStubInterface, args []string) pb.Response {
+	logger.Info("workOrderComplete")
+	if len(args) != 2 {
+		logger.Errorf("Too many parameters, expect 2, received %d", len(args))
+		return shim.Error("workOrderSubmit must include 2 arguments, workOrderID and workOrderResponse")
+	}
+
+	wo, err := t.getOrderByID(stub, args[0])
+	if err != nil {
+		logger.Errorf("Could not find the work orger by the id %s", args[0])
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `orger` to `order`

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+	err = stub.SetEvent("workOrderCompleted", eventPayload)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	logger.Info("Finished workOrderComplete")
+	return shim.Success(nil)
+}
+
+// workOrderGet - This function retrieve a work order
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `retrieve` to `retrieves`

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+func (t *WorkOrder) workOrderGet(stub shim.ChaincodeStubInterface, args []string) pb.Response {
+	logger.Info("workOrderComplete")
+	if len(args) != 1 {
+		logger.Errorf("Too many parameters, expect 1, received %d", len(args))
+		return shim.Error("workOrderSubmit must include 1 argument, workOrderID")
+	}
+
+	wo, err := t.getOrderByID(stub, args[0])
+	if err != nil {
+		logger.Errorf("Could not find the work orger by the id %s", args[0])
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `orger` to `order`

File: client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/definitions.go [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+	OBJECTTYPE   = "Receipt"
+	PAGESIZE     = 10
+	UINT64FORMAT = "%020d"
+	BYTE32FORMAT = "%-32v"
+)
+
+// Receipt data structure
+type Receipt struct {
+	WorkOrderId          string `json:"workOrgerId"`
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/definitions.go]: Change `workOrgerId` to `workOrderId`

File: client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/definitions.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+	WorkerServiceId      string `json:"workerServiceId"`
+	WorkerId             string `json:"workerId"`
+	RequesterId          string `json:"requesterId"`
+	ReceiptCreateStatus  uint64 `json:"receiptCreateStatus"`
+	WorkOrderRequestHash string `json:"workOrderRequestHash"`
+	CurrentReceiptStatus uint64 `json:"currentReceiptStatus"`
+}
+
+type ReceiptCreatedEvent struct {
+	WorkOrderId          string `json:"workOrgerId"`
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/definitions.go]: Change `workOrgerId` to `workOrderId`

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+	value, err := json.Marshal(payloadData)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	logger.Info("Finished workOrderGet")
+	return shim.Success(value)
+}
+
+// Invoke - this function simply satisfy the main requirement of chaincode
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: Change `satisfy` to `satisifies`

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/hyperledger/fabric/core/chaincode/shim"
+	pb "github.com/hyperledger/fabric/protos/peer"
+)
+
+var logger = shim.NewLogger("WorkerRegistry")
+
+// getWorkerByID - This function retrieve the worker register with its ID
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: Change `retrieve` to `retrieves`

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+	err = stub.SetEvent("workerRegistered", eventPayload)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	logger.Info("Finished workerRegister")
+	return shim.Success(nil)
+}
+
+// workerUpdate - This function set the detail of a Worker
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: Change `set` to `sets`

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+
+	err = stub.PutState(wr.WorkerID, value)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+
+	return shim.Success(value)
+}
+
+// WorkerSetStatus - This function set the status of a Worker
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: Change `set` to `sets`

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+	}
+
+	if Avalbytes == nil {
+		return shim.Error("WorkerID '" + args[0] + "' does not exist")
+	}
+
+	return shim.Success(Avalbytes)
+}
+
+// Invoke - this function simply satisfy the main requirement of chaincode
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: Change `satisfy` to `satisifies`

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+	} else if function == "workerRetrieve" {
+		return t.workerRetrieve(stub, args)
+	} else if function == "query" {
+		return t.query(stub, args)
+	}
+
+	return shim.Error("Invalid invoke function name")
+}
+
+// processAttributes - This function formalize the input attributes. It
```
Comment (danintel) [inline client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: Change `formalize` to `formalizes`

File: client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/main.go (hunk 8 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/receipt/go/main.go]: ok

File: client_sdk/avalon_client_sdk/fabric/txcommiter.py (hunk 7 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/txcommiter.py]: will change that to tx_committer.py

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go (hunk 9 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go]: this matches the API spec. Can not change that.

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go (hunk 10 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/definitions.go]: Avalon Spec, can not change this.

File: client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go (hunk 11 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/order/go/main.go]: ok

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go (hunk 20 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: ok

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go (hunk 21 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: ok

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go (hunk 22 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: ok

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go (hunk 23 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: will change it to satisfies

File: client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go (hunk 24 above)
Comment (litong01) [reply client_sdk/avalon_client_sdk/fabric/chaincode/worker/go/main.go]: ok

**Pull request merged**